### DISCUSSION TITLE
react-native-video and media source extension  API

### DISCUSSION THREAD

**Discussion by vuminhkh** [TheWidlarzGroup/react-native-video#3734] [Ideas]
### What problem are you trying to solve?

I already used react-native-video to display a camera stream from a remote URL (generated from a Home Assistant backend). It's simple and it works great, but it has some delays in this mode for live streaming. To improve the experience, I'm looking at the source code of the Frigate API, which sends chunks of data via websocket and uses the MediaSource Extension API to generate a dynamic source for the HTML video element on their web UI.

### Proposed solution

Support for the MediaSource Extension (MSE) API in react-native-video, so that a dynamic source can be generated for the video element from chunks of data streamed over a websocket (as Frigate does on their web UI), improving the live streaming experience and reducing delays.

### Alternatives considered

Displaying the camera stream from a remote URL (generated from a Home Assistant backend). This works but has some delays in live streaming mode.

### Additional context

It seems to me that react-native-video does not support the MediaSource Extension API. Is there a workaround to do the same thing?

Reference: Frigate API, which sends chunks of data via websocket and uses the MediaSource Extension API to generate a dynamic source for the HTML video element on their web UI.

**Comment by freeboub** [comment]
Looks like freegate supports rtsp. I added it 1 or 2 weeks ago (not sure it has been released). Did you try that ?

**Comment by freeboub** [comment]
Looks like freegate supports rtsp. I added it 1 or 2 weeks ago (not sure it has been released). Did you try that ?

**Reply by vuminhkh** [reply]
Hello, thanks, I'll check and test your rtsp implementation when it's released, hope that I'll have a way to get the rtsp stream via home assistant frigate integration 👍 I also tried to implement webrtc with react-native-webrtc in H264, it works great on some cameras my doorbell, my cheap wifi cameras, but on my POE annke cameras, webrtc stream freezes every 5 seconds or don't even work at all. That's why I looked into MSE in H265, it looks great, the stream is more compressed and MSE is supported by all browsers, I hope that it will be available on react native video some days.

**Comment by freeboub** [comment]
@vuminhkh There is another enhancement we can do is to add an "offsetFromLive" to bufferConfig in order to enhance latency:
See for android: https://developer.android.com/reference/kotlin/androidx/media3/exoplayer/source/DefaultMediaSourceFactory#setLiveTargetOffsetMs(long) 
For ios: https://developer.apple.com/documentation/avfoundation/avplayeritem/3227883-recommendedtimeoffsetfromlive

Should be quite easy to prototype.

I don't know well WebRTC, but it seems that RTMP can also be used for communications, This is something which can be integrated on android. It seems that rtsp and rtmp are not supported by AVPlayer ...

**Comment by freeboub** [comment]
@vuminhkh I created : https://github.com/TheWidlarzGroup/react-native-video/pull/3792
In order to be able to configure offset from live prop on android.
Let me know if it can fix your issue. I cn also do it on ios !

**Comment by vuminhkh** [comment]
@freeboub I'm not sure if your commit has anything with MSE extension, I just looked into the new props that you added to react native video, it looks like we can define playback speed / buffer, which is great because this is what's missing for me. When we check cameras recording usually I might event want to fast forward 16x. Anyway, for my little app, now I use react native video to watch recording, for streaming i'm using react native webrtc via TCP (UDP does not work with heavy WebRTC stream). I'll integrate the playback speed into my app to fast forward video recording. Thanks.

**Reply by freeboub** [reply]
Thank you for the feedback, lmet close the discussion 👍
